src/components/message/message-card/MessageCard.tsx

```diff
@@ -110,44 +126,61 @@ const MessageCard = ({ message, viewType, setViewType }: Props) => {
 		bookmarksStore.toggleMessagePin(message);
 	}, [bookmarksStore.toggleMessagePin]);
 
-	const addMessagesToExport = React.useCallback(
-		() => messagesStore.exportStore.addMessageToExport(message),
-		[messagesStore.exportStore.addMessageToExport],
-	);
-
 	const isExported = messagesStore.exportStore.isExported(message);
 
-	const rootClass = createBemBlock(
-		'message-card-wrapper',
-		isAttached ? 'attached' : null,
-		isBookmarked ? 'pinned' : null,
-		isHighlighted ? 'highlighted' : null,
-		isSoftFiltered ? 'soft-filtered' : null,
-		messagesStore.exportStore.isExport ? 'export-mode' : null,
-		isExported ? 'exported' : null,
-	);
+	const messageCardBaseProps: MessageCardBaseProps = {
+		message,
+		viewTypeConfig: viewTypesConfig,
+		hoverMessage,
+		unhoverMessage,
+		addMessagesToExport,
+		isHighlighted,
+		isSoftFiltered,
+		isExported,
+		isExpanded,
+		isExport: messagesStore.exportStore.isExport,
+		isBookmarked,
+		isAttached,
+		toogleMessagePin,
+		sortOrderItems,
+	};
 
 	return (
-		<div className={rootClass} onClick={addMessagesToExport}>
-			<MessageCardBase
-				message={message}
-				viewType={viewType}
-				setViewType={setViewType}
-				hoverMessage={hoverMessage}
-				unhoverMessage={unhoverMessage}
-				isBookmarked={isBookmarked}
-				isAttached={isAttached}
-				toogleMessagePin={toogleMessagePin}
-				sortOrderItems={sortOrderItems}
-			/>
+		<div className='messages-list__item'>
+			<MessageCardBase {...messageCardBaseProps} />
+			<MessageExpandButton isExpanded={isExpanded} setExpanded={setIsExpanded} />
 		</div>
 	);
-};
+});
 
 const RecoverableMessageCard = (props: OwnProps) => {
-	const viewTypesStore = useMessagesViewTypesStore();
-	const { viewType, setViewType } = viewTypesStore.getSavedViewType(props.message);
-	return <MessageCard {...props} viewType={viewType} setViewType={setViewType} />;
+	const viewTypesStore = useMessagesViewTypeStore();
+	const { getSavedViewType } = viewTypesStore;
+	const viewTypesConfig = props.message.parsedMessages
+		? props.message.parsedMessages.map(parsedMessage => {
+				return {
+					viewType: getSavedViewType(props.message, parsedMessage.id).viewType,
+					setViewType: getSavedViewType(props.message, parsedMessage.id).setViewType,
+				};
+		  })
+		: {
+				viewType: getSavedViewType(props.message, props.message.id).viewType,
+				setViewType: getSavedViewType(props.message, props.message.id).setViewType,
+		  };
+
+	return (
+		<StateSaver stateKey={props.message.id}>
+			{(state: boolean, stateSaver) => (
+				<MessageCard
+					{...props}
+					viewTypesConfig={viewTypesConfig}
+					stateKey={props.message.id}
+					saveState={stateSaver}
```

Comment:
i dont think MessageCars should know anything about saving state,
it should just accept viewType, and function to change it